Implements addFoundCodeAssets to find and add outputs by name.

MichealReed opened the pull request:
Adds helpers and tests for addLibraries, addAssets, addDirectories. DataAssets from addAssets and AddDirectories will not work until they are enabled by the native repo as mentioned here https://github.com/dart-lang/native/issues/2036#issuecomment-2700872475. 

Let me know what you think and I'll polish this up tomorrow!

closes #4

MichealReed commented:
So tests are going to fail for findLibraries on this because Bad state: Cannot access target architecture in dry runs.

Thinking I should add a bool for dryRun that removes this from the CodeAsset for tests. Ideas?

rainyl commented on lib/src/utils/add_assets.dart:
```diff
@@ -0,0 +1,217 @@
+  // Map the link mode preference by LinkModePreference.name
+  LinkMode linkMode;
+  List<String> libNames;
+  if (preferredMode.name == 'dynamic') {
```
Here we should use `if(preferredMode == LinkModePreference.dynamic)` to avoid hard encoding.

MichealReed replied:
That does not resolve to a type for some reason, this needs a string.

rainyl replied:
aha, I have made some changes and replaced it with `getLinkMode` copied from `native_toolchain_c`, if you have other considerations please let me know.

You can also change and polish the code directly.

rainyl commented on lib/src/utils/add_assets.dart:
```diff
@@ -0,0 +1,217 @@
+Future<List<Uri>> addAssets(
+  BuildInput input,
+  BuildOutputBuilder output,
+  Uri? outDir, {
```
`outDir` should also be optional

rainyl commented:
@MichealReed I have reviewed it and modified some code, hoping you won't mind it. 

If you have any ideas about my changes, please let me know.

- I have renamed `addLibraries` to `addCodeAssets` to keep it constant with `CodeAsset`
- `addAssets` -> `addDataAssets` to match `DataAsset`
- Replace `dynLibNames`, `staticLibNames` with `patternMap` to allow more advanced pattern match

There is also an unresolved issue in `addCodeAssets` about deciding `linkMode` of `CodeAsset`, you can take a look and try some better solutions.

MichealReed commented:
> @MichealReed I have reviewed it and modified some code, hoping you won't mind it.
> 
> If you have any ideas about my changes, please let me know.
> 
>     * I have renamed `addLibraries` to `addCodeAssets` to keep it constant with `CodeAsset`
> 
>     * `addAssets` -> `addDataAssets` to match `DataAsset`
> 
>     * Replace `dynLibNames`, `staticLibNames` with `patternMap` to allow more advanced pattern match
> 
> 
> There is also an unresolved issue in `addCodeAssets` about deciding `linkMode` of `CodeAsset`, you can take a look and try some better solutions.

I think pattern matching adds a level of complexity and it is necessary to include the extension when the input.config.code.targetOS.libraryFileName finds the proper extension via specified link mode. The previous approach would find both the static and dynamic libs and then use the one most appropriate for the given link mode.

MichealReed commented on lib/src/utils/add_assets.dart:
```diff
@@ -0,0 +1,217 @@
+  } else if (preferredMode.name == 'static') {
+    logger?.info('Using static link mode');
+    linkMode = StaticLinking();
+    libNames = staticLibNames;
```
The builder will error if you do not provide the correct asset type for the expected mode, this let me to the separation of the two.

rainyl replied:
Then I think just use `getLinkMode` is fine, I didn't meet any error about it, if it doesn't work, please attach a screenshot or some logs.

rainyl commented:
> I think pattern matching adds a level of complexity and it is not necessary to include the extension when the input.config.code.targetOS.libraryFileName finds the proper extension via specified link mode. The previous approach would find both the static and dynamic libs and then use the one most appropriate for the given link mode.

Yes, it's a little bit more complex but also more flexible, some CMake Project will create frameworks for macos or ios, and there are no extensions like `.dylib`, `.so` for libraries, just some thing like `add`, for such conditions, pattern matching can also works fine but `OS.current.dylibFileName()` fails.

MichealReed commented:
> > I think pattern matching adds a level of complexity and it is not necessary to include the extension when the input.config.code.targetOS.libraryFileName finds the proper extension via specified link mode. The previous approach would find both the static and dynamic libs and then use the one most appropriate for the given link mode.
> 
> Yes, it's a little bit more complex but also more flexible, some CMake Project will create frameworks for macos or ios, and there are no extensions like `.dylib`, `.so` for libraries, just some thing like `add`, for such conditions, pattern matching can also works fine but `OS.current.dylibFileName()` fails.
> 
> And personally I think it increases complexity significantly,
> 
> ```dart
> patternMap: {
>   RegExp(r'(lib)?add\.(so|dylib|dll)'): 'add',
> },
> ```
> 
> maybe with a more straightforward parameter name like `patternNameMap` is better?

They can always add the CodeAsset manually for edge cases, I think we should use the convention in extension OSLibraryNaming on OS { and seek to improve it there maybe?

rainyl commented:
> They can always add the CodeAsset manually for edge cases, I think we should use the convention in extension OSLibraryNaming on OS { and seek to improve it there maybe?

Allright, then I think it's fine to simplify it.

But a single `List<string> names` (like ['add', 'sub']) is okay, there is no need to separate to 2 different parameters since the library types are decided by `linkModePreference` for now.

MichealReed commented:
> > They can always add the CodeAsset manually for edge cases, I think we should use the convention in extension OSLibraryNaming on OS { and seek to improve it there maybe?
> 
> Allright, then I think it's fine to simplify it.
> 
> But a single `List<string> names` (like ['add', 'sub']) is okay, there is no need to separate to 2 different parameters since the library types are decided by `linkModePreference` for now.

Sounds good to me, I originally intended it so they could link both dynamic and static libs, but that does not seem possible as you mentioned.

> But it just a limitation of static or dynamic (it is only allowed to choose one), but frameworks on macos and ios are just packages or directories containing dynamic/static libraries without extensions, no much special things.

I will try it out at some point, but I think this depends on how they rebuild these? I ran into a lot of issues with the previous plugin cocoapods trying to rebuild libs to satisfy what flutter expected. They may be using lipo or create framework to rebuild dynamic libs into frameworks as a part of the native assets which may not be compatible with frameworks.

rainyl commented:
> But a single `List<string> names` (like ['add', 'sub']) is okay, there is no need to separate to 2 different parameters since the library types are decided by `linkModePreference` for now.

Maybe something like 

```dart
Future<List<Uri>> addCodeAssets(
  BuildInput input,
  BuildOutputBuilder output, {
  required String packageName,
  required List<String> names,
  Uri? outDir,
  Logger? logger,
}) async {
  final preferredMode = input.config.code.linkModePreference;
  final searchDir = Directory.fromUri(outDir ?? input.outputDirectory);
  final linkMode = getLinkMode(preferredMode);
  final List<Uri> foundFiles = [];

  logger?.info('Searching for libraries in ${searchDir.path}');
  logger?.info('Preferred link mode: $preferredMode');

  await for (final entity in searchDir.list(recursive: true, followLinks: false)) {
    if (entity is! File) continue;
    for (final name in names) {
      final libName = OS.current.libraryFileName(name, linkMode);
      if (entity.path.endsWith(libName)) { // this can be more elegant
        logger?.info('Found library file: ${entity.path}');
        output.assets.code.add(
          CodeAsset(
            package: packageName,
            name: '$name.dart',
            linkMode: linkMode,
            os: input.config.code.targetOS,
            file: entity.uri,
            architecture: input.config.code.targetArchitecture,
          ),
        );
        foundFiles.add(entity.uri);
        break;
      }
    }
  }
  return foundFiles;
}
```

rainyl commented:
> Sounds good to me, I originally intended it so they could link both dynamic and static libs, but that does not seem possible as you mentioned.

😄 You can work on this if you have some time today, too late for me, I am going to sleep, good luck~

> I will try it out at some point, but I think this depends on how they rebuild these? I ran into a lot of issues with the previous plugin cocoapods trying to rebuild libs to satisfy what flutter expected. They may be using lipo or create framework to rebuild dynamic libs into frameworks as a part of the native assets which may not be compatible with frameworks.

😞 Just forget it for now, in the future we can add an additional argument like `prefix`, `suffix` to allow more options.

MichealReed commented:
> 😄 You can work on this if you have some time today, too late for me, I am going to sleep, good luck~

Sure thing will brush it up, have a good rest!

MichealReed commented:
Copying the comment here for our archives.

```dart
            // TODO: the implementation of linkMode is not perfect
            //
            // The problem is, if user provides `libadd.a`, it should likely to be static,
            // but the linkmode here is fixed (decided by BuildInput) and may not exactly
            // match the type of found library (e.g., found libadd.a but linkmode is dynamic).
            // However, the `getLinkMode` function copied from `native_toolchain_c` also
            // returns `dynamic` if `linkModePreference` is `preferDynamic` (may be static)
            //
            // Maybe a better solution is inferring according to file extension, e.g.,
            // libadd.(so|dylib|dll) -> dynamic
            // libadd.(a|lib) -> static
            //
            // However, although not standard, a dynamic lib may also ends with '.a' and vice versa
            // since on unix-like OS, file type is decided by file header, maybe we can warn
            // users about this.
```

MichealReed commented:
@dcharkes are there any insights you can share given our discussion above on linking both types into a single project and use of frameworks/xcframeworks?

Also I noticed https://github.com/dart-lang/native/issues/1346. Would you like these as helpers on BuildOutput?

dcharkes commented:
> > But it just a limitation of static or dynamic (it is only allowed to choose one), but frameworks on macos and ios are just packages or directories containing dynamic/static libraries without extensions, no much special things.
> 
> I will try it out at some point, but I think this depends on how they rebuild these? I ran into a lot of issues with the previous plugin cocoapods trying to rebuild libs to satisfy what flutter expected. They may be using lipo or create framework to rebuild dynamic libs into frameworks as a part of the native assets which may not be compatible with frameworks.

Currently we don't support prepackaged xcframeworks. Flutter takes the individual dylibs, code signs them, and packages them in a framework. We've considered adding support for those, but it's low prio until we have a relevant use case. I've filed a tracking bug here: https://github.com/dart-lang/native/issues/2092

dcharkes commented:
> Also I noticed [dart-lang/native#1346](https://github.com/dart-lang/native/issues/1346). Would you like these as helpers on BuildOutput?

Yes as in the sketch. Feel free to submit a PR and migrate some of the existing `test_data/` over that have data assets. 👍

MichealReed commented:
@rainyl can we merge this version and publish a new package update? Dart native only wants the directory additions for now and that will probably land by the time data assets are actually enabled. 

I changed the library find function name to `AddFoundCodeAssets` and switched to using a map for names as the dart interface will not always match the library name in this context. For example, if the library to be found is native_add_windows_x64.dll then they must input the name as native_add_windows_x64, this resulted in the dart name being native_add_windows_x64.dart, which is wrong. There's supposed to be a fallback find mechanism for symbols like this but it was not working on Windows at least.

rainyl commented:
> Dart native only wants the directory additions for now and that will probably land by the time data assets are actually enabled.

It's reasonable to remove DataAsset-related functions for now.

> I changed the library find function name to `AddFoundCodeAssets` and switched to using a map for names as the dart interface will not always match the library name in this context.

Yes, that's why I use a `Map<..., String>` to let users to specify the name of CodeAsset.

But why you prefer `addFoundCodeAssets` than `addCodeAssets` ? Is there any extra considerations?

> can we merge this version and publish a new package update? 

I am considering to add an extension method for `BuildOutputBuilder` to simplify the calling. Will do it today if having some time, then a new version will be published.

MichealReed commented:
> But why you prefer `addFoundCodeAssets` than `addCodeAssets` ? Is there any extra considerations?

Was thinking to better indicate the find process versus adding. Fine with me whichever you prefer.

> I am considering to add an extension method for `BuildOutputBuilder` to simplify the calling. Will do it today if having some time, then a new version will be published.

Sounds good, thanks 👍

rainyl commented:
@MichealReed I have made some changes compatible with your code, added support for optional regexp matching and an extension method called `BuildOutputBuilder.findAndAddCodeAssets`, now LGTM.

Thanks for your contributions! A new release will be published soon.
